File: mktrs-spring-boot-native-java-poc/src/main/java/com/themusketeers/sbnative/users/MemoryUserService.java
/*----------------------------------------------------------------------------*/
/* Source File:   MEMORYUSERSERVICE.JAVA                                      */
/* Copyright (c), 2023 The Musketeers                                         */
/*----------------------------------------------------------------------------*/
/*-----------------------------------------------------------------------------
 History
 Jun.15/2023  COQ  File created.
 -----------------------------------------------------------------------------*/
package com.themusketeers.sbnative.users;

import com.themusketeers.sbnative.users.domain.User;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Service;

/**
 * Handles the list of users in the system. The internal representation is to
 * use a in-memory storage.
 *
 * @author COQ - Carlos Adolfo Ortiz Q.
 */
@Service
public class MemoryUserService implements UserService {
    private List<User> userList = new ArrayList<>();

    @Override
    public Boolean exists(String userId) {
        return findUserInfo(userId).isPresent();
    }

    @Override
    public User insert(User user) {
        var userToInsert = user;

        if (user.id() == null) {
            userToInsert = new User(UUID.randomUUID().toString(), user.name(), user.address());
        }

        userList.add(userToInsert);
        return userToInsert;
    }

    @Override
    public User retrieve(String userId) {
        return findUserInfo(userId).orElseGet(() -> null);
    }

    @Override
    public Boolean delete(String userId) {
        return userList.removeIf(user -> user.id().equals(userId));
    }

    @Override
    public Boolean update(User user) {
        if (exists(user.id())) {
            delete(user.id());
            insert(user);
            return true;
        }

        return false;
    }

    @Override
    public List<User> retrieveAll() {
        return userList;
    }

    @Override
    public Long count() {
        return Long.valueOf(userList.size());
    }

    private Optional<User> findUserInfo(String userId) {
        return userList
            .stream()
            .filter(user -> user.id().equals(userId))
            .findFirst();
    }
}

File: mktrs-spring-boot-native-kotlin-poc/src/main/kotlin/com/themusketeers/sbnative/service/MemoryUserService.kt
/*----------------------------------------------------------------------------*/
/* Source File:   MEMORYUSERSERVICE.KT                                        */
/* Copyright (c), 2023 The Musketeers                                         */
/*----------------------------------------------------------------------------*/
/*-----------------------------------------------------------------------------
 History
 Jun.15/2023  COQ  File created.
 -----------------------------------------------------------------------------*/
package com.themusketeers.sbnative.service

import com.themusketeers.sbnative.domain.User
import com.themusketeers.sbnative.service.intr.UserService
import java.util.UUID
import org.springframework.stereotype.Service

/**
 * Handles the list of user in the system. The internal representation is to
 * use a in-memory storage.
 *
 * @author COQ - Carlos Adolfo Ortiz Q.
 */
@Service
class MemoryUserService : UserService {
    private val userList: MutableList<User> = ArrayList()

    override fun exists(userId: String) = findUserInfo(userId).isPresent

    override fun insert(user: User): User {
        var userToInsert = user

        if (user.id == null) {
            userToInsert = User(UUID.randomUUID().toString(), user.name, user.address)
        }

        userList.add(userToInsert)
        return userToInsert
    }

    override fun retrieve(userId: String): User? = findUserInfo(userId).orElseGet { null };

    override fun delete(userId: String) = userList.removeIf { user -> user.id == userId }

    override fun update(user: User): Boolean {
        if (user.id?.let { exists(it) } == true) {
            delete(user.id)
            insert(user)
            return true
        }

        return false
    }

    override fun retrieveAll(): List<User> {
        return userList
    }

    override fun count(): Long {
        return userList.size.toLong()
    }

    private fun findUserInfo(userId: String) = userList
        .stream()
        .filter { user -> user.id == userId }
        .findFirst()
}
